objectArrayPrototype.js

```diff
@@ -0,0 +1,168 @@
+
+
+/*QUESTION 1:
+
+Creating a CAT class that will keep record of all the cats registered and 
+can return a function averageWeight() that will tell the average weight of all
+the cats that were registered.  */
+
+// Solution 1 :
+
+class CAT {
+  constructor() {
+    if (CAT.instance == null) {
+      this.logs = [];
+      CAT.instance = this;
+    } else {
+      return CAT.instance;
+    }
+  }
+
+  catStore(name, weight) {
+    this.logs.push({
+      name: name,
+      weight: weight,
+    });
+  }
+
+  averageWeight() {
+    let sum = 0;
+    for (let i = 0; i < this.logs.length; i++) {
+      sum = sum + this.logs[i].weight;
+    }
+    return sum / this.logs.length;
+  }
+
+  showAllCats() {
+    for (let i = 0; i < this.logs.length; i++) {
+      console.log(this.logs[i].name + " : " + this.logs[i].weight);
+    }
+  }
+}
+
+let garfield = new CAT();
+garfield.catStore("persia", 25);
+console.log(garfield.averageWeight());
+
+let mai = new CAT();
+mai.catStore("pozo", 15);
+console.log(mai.averageWeight());
+
+mai.showAllCats(); // shows all the cats that are in the CAT class instance.
+
+
+
+
+
+
+/*QUESTION 2:  Aim of the program :  
+
+Add all the arguments passed:
+add(1)(2)(3) => OUTPUT: 6
+add(2)(2) => OUTPUT: 4
+
+*/
+
+
+//Solution 2:
+
+function add(n) {
+  // Let the currying begin!
+  let sum = n;
+  function next(a) {
+    sum = sum + a;
+    return next;
+  }
+  next.valueOf = function () {
+    return sum;
+  };
+  return next;
+}
+
+console.log(add(1)(2)(1).valueOf());
+
+
+
+
+
+
+/* QUESTION 3:
+
+If you are calculating complex things or execute time-consuming API calls, you sometimes want to cache the results. In this case we want you to create a function wrapper, which takes a function and caches its results depending on the arguments, that were applied to the function.
+Usage example:
+var complexFunction = function(arg1, arg2) { complex calculation in here };
+var cachedFunction = cache(complexFunction);
+
+cachedFunction('foo', 'bar'); // complex function should be executed
+cachedFunction('foo', 'bar'); // complex function should not be invoked again, instead the cached result should be returned
+cachedFunction('foo', 'baz'); // should be executed, because the method wasn't invoked before with these arguments
+*/
+
+// Solution 3:
+
+function cache(complexfunction) {
+  const argumentsLogs = {};
```

Comment:
are we dealing only logs in this function?
if yes, then the function name should be cacheLogs
else variables shouldn't have suffix as logs;